src/execution_context.ts

```diff
@@ -0,0 +1,74 @@
+import * as semver from 'semver';
+import {Request, Response, NextFunction} from 'express';
+import {requiredNodeJsVersion} from './options';
+
+export const TRACE_CONTEXT_HEADER_KEY = 'X-Cloud-Trace-Context';
+export const FUNCTION_EXECUTION_ID_HEADER_KEY = 'function-execution-id';
+
+export interface ExecutionContext {
+  executionId: string;
+  traceId?: string;
+  spanId?: string;
+}
+
+const TRACE_CONTEXT_PATTERN =
+  /^(?<traceId>\w+)\/(?<spanId>\d+);o=(?<options>.+)$/;
+
+function generateExecutionId() {
+  const timestampPart = Date.now().toString(36).slice(-6);
+  const randomPart = Math.random().toString(36).slice(-6);
+  return timestampPart + randomPart;
+}
+
+let asyncLocalStorage: any;
+
+export async function executionContextMiddleware(
+  req: Request,
+  res: Response,
+  next: NextFunction
+) {
+  if (semver.lt(process.versions.node, requiredNodeJsVersion)) {
```

Comment:
I think you need to split out the logic for parsing context from the logic for configuring the asyncLocalStorage into separate middlewares.